qmi/core/transport_gpib_visa.py

```diff
@@ -0,0 +1,234 @@
+        the _device attribute is not None.
+
+        This aids in static typechecking, since whereas the type of _device is Optional[T], the result of this method is
+        guaranteed to be of type T. It is a QMI-internal bug if this property is used in case _device is None. In that
+        case, we raise an AssertionError, and we hope the users will complain to us so we can fix the bug in the
+        library.
+
+        Raises: AssertionError: in case the property is used when the underlying value of _device is None.
+
+        Returns: The value of _device, if it is not None. """
+        assert self._device is not None
+        return self._device
+
+    def __str__(self) -> str:
+        return f"QMI_VisaGpibTransport GPIB::{self._devicenr}::INSTR"
+
+    def close(self) -> None:
+        _logger.debug("Closing GPIB device nr %i", self._devicenr)
+        super().close()
+
+    def write(self, data: bytes) -> None:
+        self._check_is_open()
+
+        self._safe_device.timeout = int(self._timeout * 1000)
+        self._safe_device.write_raw(data)
+
+    def read(self, nbytes: int, timeout: Optional[float]) -> bytes:
+        """Read a specified number of bytes from the transport.
+
+        All bytes must belong to the same GPIB message.
+
+        This method blocks until the specified number of bytes are available,
+        then returns the received bytes. If timeout occurs, any partial read
+        data is discarded and QMI_TimeoutException is raised.
+
+        Parameters:
+            nbytes: Expected number of bytes to read.
+            timeout: Maximum time to wait in seconds (default: 60 seconds).
+
+        Returns:
+            Received bytes.
+
+        Raises:
+            ~qmi.core.exceptions.QMI_TimeoutException: If the timeout expires before the
+                requested number of bytes are available.
+            ~qmi.core.exceptions.QMI_EndOfInputException: If an end-of-message indicator
+                is received before the requested number of bytes is reached.
+        """
+        self._check_is_open()
+        nbuf = len(self._read_buffer)
+        while True:
+            if nbuf == nbytes:
+                # The requested number of bytes are already in the buffer. Return them immediately.
+                ret = self._read_buffer
+                self._read_buffer = bytes()
+                return ret
+
+            # GPIB requires a timeout
+            if timeout is None:
+                timeout = self._timeout
+
+            # Read buffer was of wrong length or is empty - read a new message from the instrument.
+            self._read_buffer += self._read_message(timeout)
+            nbuf = len(self._read_buffer)
+            if nbuf != nbytes:
+                _logger.debug("GPIB read buffer contained data %s" % self._read_buffer.decode())
+                self._read_buffer = bytes()
```

Comment:
If we clear the self.read_buffer, then how can there be anything left for the next read? Also, if the expected number of bytes are read, don't we also want to clear the read buffer?

Reply:
Solved this by removing the `self._read_buffer`.